packages/patternfly-4/react-core/src/components/Checkbox/Checkbox.js

```diff
@@ -1,24 +1,62 @@
 import React from 'react';
 import styles from '@patternfly/patternfly-next/components/Check/check.css';
-import { css } from '@patternfly/react-styles';
 import PropTypes from 'prop-types';
+import { css, getModifier } from '@patternfly/react-styles';
 
 const propTypes = {
-  /** additional classes added to the Checkbox */
+  /** Additional classes added to the Checkbox. */
   className: PropTypes.string,
   /** Flag to show if the Checkbox selection is valid or invalid. */
   isValid: PropTypes.bool,
   /** Flag to show if the Checkbox is disabled. */
   isDisabled: PropTypes.bool,
   /** A callback for when the Checkbox selection changes. */
-  onChange: PropTypes.func
+  onChange: PropTypes.func,
+  /** Label text of the checkbox. */
+  label: PropTypes.oneOfType([PropTypes.string, PropTypes.node, PropTypes.func]),
+  /** Id of the checkbox. */
+  id: props => {
+    if (!props.id) {
+      return new Error('CheckBox must have an id.');
+    }
+    if (typeof props.id !== 'string') {
+      return new Error(`Id must be string. Received ${typeof props.id}.`);
+    }
+    return null;
+  },
+  /** Aria-label of the checkbox. */
+  'aria-label': props => {
+    if (typeof props.label !== 'string' && !props['aria-label']) {
+      return new Error('CheckBox must have an aria-label, when label is not string.');
+    }
+    if (props['aria-label'] && typeof props.id !== 'string') {
+      return new Error(`Aria-label must be string. Received ${typeof props.id}.`);
+    }
+    return null;
+  }
 };
 
 const defaultProps = {
   className: '',
   isValid: true,
   isDisabled: false,
-  onChange: () => undefined
+  onChange: () => undefined,
+  label: undefined,
+  id: null,
+  'aria-label': null
+};
+
+const renderLabel = (label, checkId, isDisabled) => {
+  if (typeof label === 'function') {
+    return label();
+  } else if (typeof label === 'object') {
+    return label;
```

Comment:
This would not work if someone actually passed an object here since react cannot render objects. A better check would be `React.isValidElement(label)`, but that would break styles if I was trying to use react-intl. It is a react node that renders a string.

```
<Checkbox 
  label={<FormattedMessage defaultMessage="Some Label" id="someId" />}
/>
```
That will render a string that isn't wrapped by the label. For comparrison this would work correctly in material-ui.

Reply:
> This would not work if someone actually passed an object here since react cannot render objects.

To be absolutely sure that the object is react node i think there should be something like: 
```javascript
typeof label === 'object' && label.$$type === 'React.Node'
``` 
(JS qualifies react nodes ad objects, so if you are not using something like flow types from typescript you need the 'object' compassion)

Reply:
`React.isValidElemen`t does this in a more maintainable way: https://github.com/facebook/react/blob/master/packages/react/src/ReactElement.js#L366 This will allow us to not have to make changes based on React internals. the `$$` prefix is React telling consumers that is should be considered private.

Reply:
I still do not believe this is resolved. While I mentioned using the check it still does not fix the fact that if I want to render an element which in turn renders only a string. This would render it without being wrapped in our styled `<label>` element.